### PR TITLE
ELEMENTS-1275: selectivity dropdown close

### PR DESCRIPTION
**Pull request by richardsd**

### CODE REVIEW

**Comment by richardsd** [comment]
selectivity_close_on_scroll_example

File: ui/widgets/nuxeo-selectivity.js [hunk 1]
```diff
@@ -7214,16 +7214,46 @@ typedArrayTags[weakMapTag] = false;
 
       // scope styles
       this.scopeSubtree(this.$.input, true);
+
+      // use observer to hide dropdown if not visible in the viewport
+      this._visibilityObserver = new IntersectionObserver((entries) => {
+        if (entries && entries.length > 0) {
+          const entry = entries[0];  // there should be only one entry here
+          if (!entry.isVisible) {
+            this._selectivity.close();
```
Comment (Gabez0r) [inline ui/widgets/nuxeo-selectivity.js:7223]: nit: `this._closeSelectivityDropdown()` for consistency.

File: ui/widgets/nuxeo-selectivity.js [hunk 2]
```diff
@@ -7214,16 +7214,46 @@ typedArrayTags[weakMapTag] = false;
...
+      }, {
+        rootMargin: '0px',
+        threshold: 1.0
+      });
+      this._visibilityObserver.observe(this);
+
+      // attach event listener to scroller parent to hide dropdown
+      this._scrollParent = this._getScrollParent(this);
+      if (this._scrollParent) {
+        this._scrollParent.addEventListener('scroll', this._closeSelectivityDropdown.bind(this));
```
Comment (Gabez0r) [inline ui/widgets/nuxeo-selectivity.js:7235]: With this in place, do we still need the IntersectionObserver?

Comment (richardsd) [reply ui/widgets/nuxeo-selectivity.js:7235]: No, we don't need the `IntersectionObserver` if we go with the scrollable parent approach. I kept both solutions here to spark the discussion.

Comment (Gabez0r) [reply ui/widgets/nuxeo-selectivity.js:7235]: Aah, sorry. Thank you. Can you please add that to the description of the PR? Didn't notice we had two alternatives here.

Comment (richardsd) [reply ui/widgets/nuxeo-selectivity.js:7221]: Just removed the use of the IntersectionObserver, since we're going with the scrollable parent solution.

**Comment by richardsd** [comment]
Pipeline at https://qa.nuxeo.org/jenkins/job/web-ui-pipeline-rdias/26/ 🕐

File: ui/widgets/nuxeo-selectivity.js [hunk 3]
```diff
@@ -7214,16 +7214,27 @@ typedArrayTags[weakMapTag] = false;
 
       // scope styles
       this.scopeSubtree(this.$.input, true);
+
+      // attach event listener to scroller parent to hide dropdown
+      this._scrollParent = this._getScrollParent(this);
```
Comment (Gabez0r) [inline ui/widgets/nuxeo-selectivity.js:7219]: I guess we don't need the to pass `this` as a parameter here?

File: ui/widgets/nuxeo-selectivity.js [hunk 4]
```diff
@@ -7214,16 +7214,27 @@ typedArrayTags[weakMapTag] = false;
 
       // scope styles
       this.scopeSubtree(this.$.input, true);
+
+      // attach event listener to scroller parent to hide dropdown
```
Comment (Gabez0r) [inline ui/widgets/nuxeo-selectivity.js:7218]: Can you please make it a bit more explicit here that we are closing the dropdown whenever we scroll on the parent? So we can later understand that this was intended.

**Review by Gabez0r — APPROVED** [review]
LGTM. I think we could improve this by only closing the dropdown when it goes out range and by making sure it follows the input. I think we could achieve this by calling `positionDropdown` when we scroll. But we can leave this for a followup ticket. WDYT?

File: ui/widgets/nuxeo-selectivity.js (hunk 3 above)
Comment (richardsd) [reply ui/widgets/nuxeo-selectivity.js:7219]: Nope, you're right. I'll remove it.

**Comment by richardsd** [comment]
> LGTM. I think we could improve this by only closing the dropdown when it goes out range and by making sure it follows the input. I think we could achieve this by calling `positionDropdown` when we scroll. But we can leave this for a followup ticket. WDYT?

I am totally fine with that solution as well. Since we already know the scroller parent, we just need to call the `positionDropdown` on the selectivity (tested it, and it works).

**Comment by richardsd** [comment]
@Gabez0r is this the behavior we were discussing? We can do it with the `IntersectionObserver` like we had before.
selectivity_close_on_scroll_example_v2

**Comment by richardsd** [comment]
Pipeline at https://qa.nuxeo.org/jenkins/job/web-ui-pipeline-rdias/28/

File: ui/widgets/nuxeo-selectivity.js [hunk 5]
```diff
@@ -7373,6 +7400,38 @@ typedArrayTags[weakMapTag] = false;
...
+      const overflowRegex = /(auto|scroll)/;
+
+      if (style.position !== 'fixed') {
+        let parent = this;
+        while (parent) {
+          if (parent.parentElement) {
+            parent = parent.parentElement;
+          } else if (parent.getRootNode()) {
+            parent = parent.getRootNode();
+            if (parent) {
```
Comment (mnixo) [inline ui/widgets/nuxeo-selectivity.js:7417]: Isn't the last `if` going to evaluate as `true` every time? You already checked for `parent` in the previous `if`, when it was still `parent.getRootNode()`. 
The way I see it:
```
} else if (a) {
  b = a;
  if (b) {
```
If you checked for `a` before, why check for `b` when `b` is `a`?

**Review by mnixo — APPROVED** [review]

File: ui/widgets/nuxeo-selectivity.js (hunk 5 above)
Comment (richardsd) [reply ui/widgets/nuxeo-selectivity.js:7417]: You're right Miguel. This is what happens when you start refactoring code and don't clean up stuff properly. Thanks, I'll update that.

**Review by mnixo — APPROVED** [review]

**Review by Gabez0r — APPROVED** [review]

**Pull request merged**